server/src/test/java/org/elasticsearch/index/snapshots/blobstore/SlicedInputStreamTests.java

```diff
@@ -78,6 +79,122 @@ protected InputStream openSlice(int slice) throws IOException {
+        SlicedInputStream input = new SlicedInputStream(slices) {
+            @Override
+            protected InputStream openSlice(int slice) throws IOException {
+                final int sliceOffset = slice * sliceSize;
+                final int length = slice == slices - 1 ? bytes.length - sliceOffset : sliceSize;
+                final var stream = new CheckClosedInputStream(new ByteArrayInputStream(bytes, sliceOffset, length));
+                streamsOpened.add(stream);
+                return stream;
+            }
+        };
+
+        // Read up to a random point
+        final int mark = randomIntBetween(0, bytes.length);
+        if (mark > 0) {
+            final var bytesReadUntilMark = new byte[mark];
+            input.readNBytes(bytesReadUntilMark, 0, mark);
+            final var expectedBytesUntilMark = new ByteArrayInputStream(bytes, 0, mark).readAllBytes();
+            assertArrayEquals(expectedBytesUntilMark, bytesReadUntilMark);
+        }
+
+        // Reset should throw since there is no mark
+        expectThrows(IOException.class, input::reset);
+
+        // Mark
+        input.mark(Integer.MAX_VALUE);
+
+        // Read up to another random point
+        final int moreBytes = randomIntBetween(0, bytes.length - mark);
+        if (moreBytes > 0) {
+            final var moreBytesRead = new byte[moreBytes];
+            input.readNBytes(moreBytesRead, 0, moreBytes);
+            final var expectedMoreBytes = new ByteArrayInputStream(bytes, mark, moreBytes).readAllBytes();
+            assertArrayEquals(expectedMoreBytes, moreBytesRead);
+        }
+
+        // Reset
+        input.reset();
+
+        // Read all remaining bytes, which should be the bytes from mark up to the end
+        final int remainingBytes = bytes.length - mark;
+        if (remainingBytes > 0) {
+            final var remainingBytesRead = new byte[remainingBytes];
+            input.readNBytes(remainingBytesRead, 0, remainingBytes);
+            final var expectedRemainingBytes = new ByteArrayInputStream(bytes, mark, remainingBytes).readAllBytes();
+            assertArrayEquals(expectedRemainingBytes, remainingBytesRead);
+        }
+
+        // Confirm we reached the end and close the stream
+        assertThat(input.read(), equalTo(-1));
+        input.close();
+        streamsOpened.forEach(stream -> assertTrue(stream.closed));
+    }
+
+    public void testMarkResetClosedStream() throws IOException {
+        final int slices = randomIntBetween(1, 20);
+        SlicedInputStream input = new SlicedInputStream(slices) {
+            @Override
+            protected InputStream openSlice(int slice) throws IOException {
+                return new ByteArrayInputStream(new byte[] { 0 }, 0, 1);
+            }
+        };
+
+        input.skip(randomIntBetween(1, slices));
+        input.mark(Integer.MAX_VALUE);
+        input.close();
+        assertThat(input.read(), equalTo(-1));
```

Comment:
This is slightly strange behavior but I suppose unmodified. I'd add a comment...